block/azure/adapter.go

```diff
@@ -0,0 +1,505 @@
+)
+
+const (
+	BlockstoreType          = "wasb"
+	sizeSuffix              = "_size"
+	idSuffix                = "_id"
+	_1MiB                   = 1024 * 1024
+	defaultMaxRetryRequests = 0
+)
+
+type Adapter struct {
+	ctx            context.Context
+	p              pipeline.Pipeline
+	accountName    string
+	configurations configurations
+}
+
+type configurations struct {
+	retryReaderOptions azblob.RetryReaderOptions
+}
+
+func NewAdapter(pipeline pipeline.Pipeline, accountName string, opts ...func(a *Adapter)) *Adapter {
+	a := &Adapter{
+		ctx:            context.Background(),
+		accountName:    accountName,
+		p:              pipeline,
+		configurations: configurations{retryReaderOptions: azblob.RetryReaderOptions{MaxRetryRequests: defaultMaxRetryRequests}},
+	}
+	for _, opt := range opts {
+		opt(a)
+	}
+	return a
+}
+
+func (a *Adapter) WithContext(ctx context.Context) block.Adapter {
+	return &Adapter{
+		p:           a.p,
+		accountName: a.accountName,
+		ctx:         ctx,
+	}
+}
+
+func (a *Adapter) log() logging.Logger {
+	return logging.FromContext(a.ctx)
+}
+
+func resolveNamespace(obj block.ObjectPointer) (block.QualifiedKey, error) {
+	qualifiedKey, err := block.ResolveNamespace(obj.StorageNamespace, obj.Identifier)
+	if err != nil {
+		return qualifiedKey, err
+	}
+	if qualifiedKey.StorageType != block.StorageTypeAzure {
+		return qualifiedKey, block.ErrInvalidNamespace
+	}
+	return qualifiedKey, nil
+}
+
+func resolveNamespacePrefix(lsOpts block.WalkOpts) (block.QualifiedPrefix, error) {
+	qualifiedPrefix, err := block.ResolveNamespacePrefix(lsOpts.StorageNamespace, lsOpts.Prefix)
+	if err != nil {
+		return qualifiedPrefix, err
+	}
+	if qualifiedPrefix.StorageType != block.StorageTypeAzure {
+		return qualifiedPrefix, block.ErrInvalidNamespace
+	}
+	return qualifiedPrefix, nil
+}
+
+func (a *Adapter) GenerateInventory(ctx context.Context, logger logging.Logger, inventoryURL string, shouldSort bool, prefixes []string) (block.Inventory, error) {
+	return nil, fmt.Errorf("inventory %w", ErrNotImplemented)
+}
+
+func (a *Adapter) getContainerURL(containerName string) azblob.ContainerURL {
+	URL, _ := url.Parse(
+		fmt.Sprintf("https://%s.blob.core.windows.net/%s", a.accountName, containerName))
+
+	return azblob.NewContainerURL(*URL, a.p)
+}
+
+func (a *Adapter) getBlobURL(containerName, fileName string) azblob.BlobURL {
+	containerURL := a.getContainerURL(containerName)
+
+	return containerURL.NewBlobURL(fileName)
+}
+
+func (a *Adapter) getBlockBlobURL(containerName, fileName string) azblob.BlockBlobURL {
+	containerURL := a.getContainerURL(containerName)
+
+	return containerURL.NewBlockBlobURL(fileName)
+}
+
+func (a *Adapter) getIDURL(containerName, fileName string) azblob.BlockBlobURL {
+	return a.getBlockBlobURL(containerName, fileName+idSuffix)
+}
+
+func (a *Adapter) getSizeURL(containerName, fileName string) azblob.BlockBlobURL {
+	return a.getBlockBlobURL(containerName, fileName+sizeSuffix)
+}
```

Comment:
Think we just need one function that builds the container URL object and not a lot of helpers that will continue to call the base function that will parse the URL again and again